File: java/omnia/data/structure/mutable/MaskingList.java
package omnia.data.structure.mutable;

import java.util.Iterator;
import java.util.OptionalInt;
import java.util.stream.Stream;

class MaskingList<E> implements MutableList<E> {

  private final java.util.List<E> javaList;

  MaskingList(java.util.List<E> javaList) {
    this.javaList = javaList;
  }

  @Override
  public void insertAt(int index, E element) {
    javaList.add(index, element);
  }

  @Override
  public void removeAt(int index) {
    javaList.remove(index);
  }

  @Override
  public void replaceAt(int index, E element) {
    javaList.set(index, element);
  }

  @Override
  public E itemAt(int index) {
    return javaList.get(index);
  }

  @Override
  public OptionalInt indexOf(Object element) {
    int index = javaList.indexOf(element);
    return index < 0 ? OptionalInt.empty() : OptionalInt.of(index);
  }

  @Override
  public void add(E element) {
    javaList.add(element);
  }

  @Override
  public boolean remove(E element) {
    return javaList.remove(element);
  }

  @Override
  public void clear() {
    javaList.clear();
  }

  @Override
  public Iterator<E> iterator() {
    return javaList.iterator();
  }

  @Override
  public boolean contains(Object element) {
    return javaList.contains(element);
  }

  @Override
  public boolean isPopulated() {
    return !javaList.isEmpty();
  }

  @Override
  public int count() {
    return javaList.size();
  }

  @Override
  public Stream<E> stream() {
    return javaList.stream();
  }
}

File: src/omnia/data/structure/mutable/MaskingList.kt
package omnia.data.structure.mutable

import omnia.data.structure.Collection

open class MaskingList<E : Any>(private val kotlinList: kotlin.collections.MutableList<E>) :
  MutableList<E> {

  override fun insertAt(index: Int, item: E) {
    kotlinList.add(index, item)
  }

  override fun removeAt(index: Int): E {
    return kotlinList.removeAt(index)
  }

  override fun replaceAt(index: Int, item: E): E {
    return kotlinList.set(index, item)
  }

  override fun itemAt(index: Int): E {
    return kotlinList.get(index)
  }

  override fun indexOf(item: Any?): Int? {
    val index: Int = kotlinList.indexOf(item)
    return if (index < 0) null else index
  }

  override fun add(item: E) {
    kotlinList.add(item)
  }

  override fun addAll(items: Collection<out E>) {
    kotlinList.addAll(items)
  }

  override fun removeUnknownTyped(item: Any?): Boolean {
    return kotlinList.remove(item)
  }

  override fun clear() {
    kotlinList.clear()
  }

  override fun iterator(): MutableIterator<E> {
    return kotlinList.iterator()
  }

  override fun containsUnknownTyped(item: Any?): Boolean {
    return kotlinList.contains(item)
  }

  override val isPopulated: Boolean
    get() = kotlinList.isNotEmpty()

  override fun count(): Int {
    return kotlinList.size
  }
}